File: itemerness-core/src/main/kotlin/com/iroselle/itemerness/core/presentation/Formatting.kt
package com.iroselle.itemerness.core.presentation

import com.iroselle.itemerness.api.BooleanDataValue
import com.iroselle.itemerness.api.CompoundDataValue
import com.iroselle.itemerness.api.DecimalDataValue
import com.iroselle.itemerness.api.IntegerDataValue
import com.iroselle.itemerness.api.ItemDataValue
import com.iroselle.itemerness.api.ItemKey
import com.iroselle.itemerness.api.ListDataValue
import com.iroselle.itemerness.api.LongDataValue
import com.iroselle.itemerness.api.NamespacedKeyDataValue
import com.iroselle.itemerness.api.StringDataValue
import com.iroselle.itemerness.api.UuidDataValue
import java.math.BigDecimal
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale

internal class MessageResolver(
    private val catalog: PresentationCatalogSnapshot,
    requestedLocale: String,
) {
    val locale: String = if (requestedLocale in catalog.locales) requestedLocale else catalog.defaultLocale
    private val chain: List<LocaleSource> = buildList {
        val visited = HashSet<String>()
        var current: String? = locale
        while (current != null && visited.add(current)) {
            val source = catalog.locales[current] ?: break
            add(source)
            current = source.fallback
        }
        if (none { it.locale == catalog.defaultLocale }) {
            catalog.locales[catalog.defaultLocale]?.let(::add)
        }
    }

    fun resolve(key: String, arguments: Map<String, String> = emptyMap()): String? {
        val raw = chain.firstNotNullOfOrNull { it.messages[key] } ?: return null
        if (arguments.isEmpty()) return raw
        return PLACEHOLDER_PATTERN.replace(raw) { match -> arguments[match.groupValues[1]] ?: match.value }
    }

    private companion object {
        val PLACEHOLDER_PATTERN = Regex("\\{([a-zA-Z0-9_.-]+)}")
    }
}

internal class ValueFormatter(
    private val catalog: PresentationCatalogSnapshot,
    private val messages: MessageResolver,
) {
    fun format(value: ItemDataValue, format: ItemKey? = null): Result<String> = runCatching {
        if (format == null) defaultFormat(value) else applyFormat(value, format, LinkedHashSet())
    }

    private fun applyFormat(value: ItemDataValue, id: ItemKey, active: MutableSet<ItemKey>): String {
        check(active.add(id)) { "Formatter recursion detected at $id" }
        try {
            return when (val format = requireNotNull(catalog.formats[id]) { "Unknown formatter $id" }) {
                is FormatSource.IntegerFormat -> {
                    val number = when (value) {
                        is IntegerDataValue -> value.value.toLong()
                        is LongDataValue -> value.value
                        else -> error("Formatter $id requires integer data")
                    }
                    decimalFormat(format.pattern).format(number)
                }

                is FormatSource.DecimalFormat -> {
                    val number = when (value) {
                        is DecimalDataValue -> value.value
                        is IntegerDataValue -> value.value.toDouble()
                        is LongDataValue -> value.value.toDouble()
                        else -> error("Formatter $id requires decimal data")
                    }
                    val scaled = number * format.multiply
                    require(scaled.isFinite()) { "Formatter $id produced a non-finite number" }
                    val suffix = format.suffixMessage?.let {
                        requireNotNull(messages.resolve(it)) { "Missing message $it" }
                    }.orEmpty()
                    decimalFormat(format.pattern).format(scaled) + suffix
                }

                is FormatSource.BooleanFormat -> {
                    val boolean = (value as? BooleanDataValue)?.value ?: error("Formatter $id requires boolean data")
                    val key = if (boolean) format.trueMessage else format.falseMessage
                    requireNotNull(messages.resolve(key)) { "Missing message $key" }
                }

                is FormatSource.NamespacedKeyFormat -> {
                    val key = (value as? NamespacedKeyDataValue)?.value ?: error("Formatter $id requires namespaced-key data")
                    when (format.mode) {
                        NamespacedKeyFormatMode.PATH -> key.value
                        NamespacedKeyFormatMode.MESSAGE -> {
                            val pattern = requireNotNull(format.messagePattern) { "Formatter $id has no message pattern" }
                            val messageKey = pattern
                                .replace("{namespace}", key.namespace)
                                .replace("{path}", key.value)
                            messages.resolve(messageKey) ?: when (format.missingValue) {
                                MissingKeyValue.PATH -> key.value
                                MissingKeyValue.FULL_KEY -> key.toString()
                                MissingKeyValue.ERROR -> error("Missing value message $messageKey")
                            }
                        }
                    }
                }

                is FormatSource.ListFormat -> {
                    val list = (value as? ListDataValue)?.values ?: error("Formatter $id requires list data")
                    val elementFormat = ItemKey.parse(format.elementFormat)
                    val separator = requireNotNull(messages.resolve(format.separatorMessage)) {
                        "Missing message ${format.separatorMessage}"
                    }
                    list.joinToString(separator) { applyFormat(it, elementFormat, active) }
                }
            }
        } finally {
            active.remove(id)
        }
    }

    private fun defaultFormat(value: ItemDataValue): String = when (value) {
        is BooleanDataValue -> value.value.toString()
        is IntegerDataValue -> value.value.toString()
        is LongDataValue -> value.value.toString()
        is DecimalDataValue -> BigDecimal.valueOf(value.value).stripTrailingZeros().toPlainString()
        is StringDataValue -> value.value
        is UuidDataValue -> value.value.toString()
        is NamespacedKeyDataValue -> value.value.toString()
        is ListDataValue -> value.values.joinToString(", ") { defaultFormat(it) }
        is CompoundDataValue -> value.entries.entries.joinToString(", ") { (key, nested) -> "$key=${defaultFormat(nested)}" }
    }

    private fun decimalFormat(pattern: String): DecimalFormat {
        val locale = Locale.forLanguageTag(messages.locale.replace('_', '-'))
        return DecimalFormat(pattern, DecimalFormatSymbols.getInstance(locale)).apply {
            isParseBigDecimal = true
            roundingMode = java.math.RoundingMode.HALF_UP
        }
    }
}

internal fun comparePresentationValues(left: ItemDataValue?, right: ItemDataValue?): Int? {
    if (left == null || right == null) return null
    val leftNumber = numericValue(left)
    val rightNumber = numericValue(right)
    if (leftNumber != null && rightNumber != null) return leftNumber.compareTo(rightNumber)
    return when {
        left is StringDataValue && right is StringDataValue -> left.value.compareTo(right.value)
        left is NamespacedKeyDataValue && right is NamespacedKeyDataValue -> left.value.compareTo(right.value)
        left is UuidDataValue && right is UuidDataValue -> left.value.compareTo(right.value)
        left is BooleanDataValue && right is BooleanDataValue -> left.value.compareTo(right.value)
        else -> null
    }
}

private fun numericValue(value: ItemDataValue): BigDecimal? = when (value) {
    is IntegerDataValue -> BigDecimal.valueOf(value.value.toLong())
    is LongDataValue -> BigDecimal.valueOf(value.value)
    is DecimalDataValue -> BigDecimal.valueOf(value.value)
    else -> null
}

File: editor/packages/mc-render/src/formatting.ts
import {
    namespacedIdSchema,
    type DataTypeNode,
    type DataValue,
    type FormatNode,
} from "@itemerness/protocol";
import {
    formatDefaultDecimal,
    formatJavaNumber,
    LocalFormatError,
} from "./decimalFormat.js";

export interface FormatMessages {
    readonly effectiveLocale: string;
    resolve(key: string): string | null;
}

/** Formatting.kt's value and failure semantics, shared by every local content block. */
export class LocalValueFormatter {
    private readonly formats: ReadonlyMap<string, FormatNode>;

    constructor(
        formats: readonly FormatNode[],
        private readonly messages: FormatMessages,
    ) {
        this.formats = new Map(formats.map((format) => [format.id, format]));
        if (this.formats.size !== formats.length)
            throw new LocalFormatError("Formatter identifiers must be unique");
        const visited = new Set<string>();
        const active = new Set<string>();
        const visit = (id: string) => {
            if (active.has(id))
                throw new LocalFormatError(
                    `Formatter recursion detected at ${id}`,
                );
            if (visited.has(id)) return;
            const format = this.formats.get(id);
            if (!format) throw new LocalFormatError(`Unknown formatter ${id}`);
            active.add(id);
            if (format.kind === "list") visit(format.elementFormat);
            active.delete(id);
            visited.add(id);
        };
        formats.forEach((format) => visit(format.id));
    }

    format(
        value: DataValue,
        id: string | null = null,
        type?: DataTypeNode,
    ): string {
        return id === null
            ? this.defaultFormat(value, type)
            : this.apply(value, id, new Set(), type);
    }

    requiredMessage(key: string): string {
        const value = this.messages.resolve(key);
        if (value === null)
            throw new LocalFormatError(`Missing message ${key}`);
        return value;
    }

    private apply(
        value: DataValue,
        id: string,
        active: Set<string>,
        type?: DataTypeNode,
    ): string {
        if (active.has(id))
            throw new LocalFormatError(`Formatter recursion detected at ${id}`);
        const format = this.formats.get(id);
        if (!format) throw new LocalFormatError(`Unknown formatter ${id}`);
        active.add(id);
        try {
            switch (format.kind) {
                case "integer":
                    if (
                        value.kind !== "integer" ||
                        (type &&
                            type.kind !== "integer" &&
                            type.kind !== "long")
                    )
                        throw new LocalFormatError(
                            `Formatter ${id} requires integer data`,
                        );
                    return formatJavaNumber(
                        value.value,
                        "integer",
                        format.pattern,
                        this.messages.effectiveLocale,
                    );
                case "decimal": {
                    if (
                        (value.kind !== "integer" &&
                            value.kind !== "decimal") ||
                        (type &&
                            type.kind !== "integer" &&
                            type.kind !== "long" &&
                            type.kind !== "decimal")
                    )
                        throw new LocalFormatError(
                            `Formatter ${id} requires decimal data`,
                        );
                    const number = formatJavaNumber(
                        value.value,
                        "decimal",
                        format.pattern,
                        this.messages.effectiveLocale,
                        format.multiply,
                    );
                    return (
                        number +
                        (format.suffixMessage === null
                            ? ""
                            : this.requiredMessage(format.suffixMessage))
                    );
                }
                case "boolean":
                    if (value.kind !== "boolean")
                        throw new LocalFormatError(
                            `Formatter ${id} requires boolean data`,
                        );
                    return this.requiredMessage(
                        value.value ? format.trueMessage : format.falseMessage,
                    );
                case "namespacedKey": {
                    if (
                        value.kind !== "string" ||
                        (type && type.kind !== "namespacedKey") ||
                        !namespacedIdSchema.safeParse(value.value).success
                    )
                        throw new LocalFormatError(
                            `Formatter ${id} requires namespaced-key data`,
                        );
                    const [namespace, path] = value.value.split(":") as [
                        string,
                        string,
                    ];
                    if (format.mode === "PATH") return path;
                    if (format.messagePattern === null)
                        throw new LocalFormatError(
                            `Formatter ${id} has no message pattern`,
                        );
                    const key = format.messagePattern
                        .replaceAll("{namespace}", namespace)
                        .replaceAll("{path}", path);
                    const message = this.messages.resolve(key);
                    if (message !== null) return message;
                    if (format.missingValue === "PATH") return path;
                    if (format.missingValue === "FULL_KEY") return value.value;
                    throw new LocalFormatError(`Missing value message ${key}`);
                }
                case "list": {
                    if (value.kind !== "list")
                        throw new LocalFormatError(
                            `Formatter ${id} requires list data`,
                        );
                    const separator = this.requiredMessage(
                        format.separatorMessage,
                    );
                    if (!this.formats.has(format.elementFormat))
                        throw new LocalFormatError(
                            `Unknown formatter ${format.elementFormat}`,
                        );
                    return value.values
                        .map((entry) =>
                            this.apply(
                                entry,
                                format.elementFormat,
                                active,
                                type?.kind === "list"
                                    ? type.element
                                    : undefined,
                            ),
                        )
                        .join(separator);
                }
            }
        } finally {
            active.delete(id);
        }
    }

    private defaultFormat(value: DataValue, type?: DataTypeNode): string {
        switch (value.kind) {
            case "null":
                throw new LocalFormatError(
                    "Null cannot be formatted as an item data value",
                );
            case "boolean":
                return String(value.value);
            case "integer":
                return type?.kind === "decimal"
                    ? formatDefaultDecimal(value.value)
                    : BigInt(value.value).toString();
            case "decimal":
                return formatDefaultDecimal(value.value);
            case "string":
                return type?.kind === "uuid"
                    ? value.value.toLowerCase()
                    : value.value;
            case "list":
                return value.values
                    .map((entry) =>
                        this.defaultFormat(
                            entry,
                            type?.kind === "list" ? type.element : undefined,
                        ),
                    )
                    .join(", ");
            case "compound":
                return Object.keys(value.entries)
                    .sort()
                    .map(
                        (key) =>
                            `${key}=${this.defaultFormat(value.entries[key]!, type?.kind === "compound" ? type.fields?.find((field) => field.name === key)?.type : undefined)}`,
                    )
                    .join(", ");
        }
    }
}
